pyproject.toml

```diff
@@ -46,17 +46,10 @@ pytest-rerunfailures = "^11.1.2"
 coverage = "^7.2.6"
 
 [tool.poetry.group.lint.dependencies]
+ruff = "^0.0.278"
 black = "^23.3.0"
 coverage = "^7.2.6"
-flake8 = "^6.0.0"
-flake8-annotations = "^3.0.1"
-flake8-bugbear = "^23.5.9"
-flake8-tidy-imports = "^4.8.0"
-flake8-pep585 = {version = "^0.1.7", python = ">=3.9"}
-flake8-new-union-types = "^0.4.1"
-flake8-future-annotations = "^1.1.0"
```

Comment:
This is really questionable. Instead of implementing the own version of flake8's API for being backwards-compatible with all plugins, they ended up rewriting every popular plugin in rust. This gives us quite outdated version of packages and may only be worse, if `ruff` will not have strong maintaining.

Although, I am okay using it now, because it can address all our needs and we are not planning to add more flake8 plugins in the future.

Reply:
It is true that Ruff could use a plugin system, and it is [already in discussion](https://github.com/astral-sh/ruff/issues/283).

However, the main point of ruff is to be really fast, to do that, first party support was preferred, especially when you consider that ruff is a rust library (it's unclear whether the plugins to ruff would then be written in python, or rust, if in python, that would likely greatly slow it down, if in rust, it might be pretty annoying to write for people not too familiar). 

IMO ruff did a really good job at re-implementing essentially all of the popular flake8 extensions, or at least providing equivalent or near equivalent (and often better)  alternatives to them form different rule-sets. I have been working on migrating various other projects of mine that use much more flake8 extensions than we do in mcstatus, and so far, ruff had support, or alternatives for all of them.

That said, there are of course a lot of much less commonly used flake8 extensions, or even completely custom ones that enforce a very specific thing unique to a codebase, and for that, ruff might not currently be the best option (at least until the plugin system does get added). However as you said, we don't really need anything too custom in mcstatus, and with everything Ruff currently supports, it should be more than enough.

---

You did also mention a point on backwards compatibility, and how Ruff will have to actively keep track of the plugins that it "carried over". I'd like to talk a bit on that, as it's pretty important to explain how Ruff handles this.

Generally, the burden of maintaining any of the ported rules come to Ruff's maintainers, and they do accept this, and aim to mostly remain compatible. 

That said, Ruff is already not 100% compatible with the ported extensions, and that has various reasons. The biggest one is that many of the ported extensions had some overlap in the rules, and so Ruff only kept one of them, sometimes modifying that one to merge the behaviors from both extensions, when that made sense. 

The other reason is that Ruff addresses the bugs reported to them directly, so in many cases, the ported rules were changed to fix a bug, or even introduce a new subfeature for that rule. That means even though a rule might have the same error code, it's not always the same rule in what it's enforcing.

This is because Ruff is not trying to be a mere 1:1 replica of flake8 and the extension that were carried over. It's very important to understand that Ruff is it's own thing, and the rules, included those originally migrated from other flake8 extensions are treated as first party rules, with their own, potentially different implementation from the one in the ported extension.

Ruff might even eventually add new rules under an existing ruleset of some extension, if it makes contextual sense, say there's a new PEP about annotations, ruff could absolutely implement a rule for it under the `flake8-annotations` rule-set, even though the original extension didn't implement it.

They generally try not to do this, to avoid any potential rule clashes if the extension later does implement it, but it's not out of question, and especially if ruff at some point becomes more popular than flake8, which it is on a pretty good track for, it's the flake8 extensions that might want to start confirming to ruff's rules, rather than the other way around. Though right now, new custom rules are being introduced purely under the `RUF` category, but they are mostly pretty general rules, applicable everywhere, and not something that could potentially belong udner any of the extensions rulesets.